plugins/json/json.cxx

```diff
@@ -0,0 +1,391 @@
+	 *
+	 * \param json_data A json data to which we write a scalar data
+	 * \param reference A reference to a scalar datatype
+	 * \param logger PDI's logger instance
+	 */
+	void write_scalar_to_json(nlohmann::json& json_data, Ref_r reference, Logger& logger)
+	{
+		auto scalar_type = std::dynamic_pointer_cast<const Scalar_datatype>(reference.type());
+		if (scalar_type->kind() == Scalar_kind::UNSIGNED) {
+			if (scalar_type->buffersize() == 1L) {
+				json_data = std::string(1, reference.scalar_value<char>());
+			} else if (scalar_type->buffersize() == 2L) {
+				json_data = reference.scalar_value<uint16_t>();
+			} else if (scalar_type->buffersize() == 4L) {
+				json_data = reference.scalar_value<uint32_t>();
+			} else if (scalar_type->buffersize() == 8L) {
+				json_data = reference.scalar_value<uint64_t>();
+			} else {
+				throw Type_error{"Unknown size of unsigned integer datatype"};
+			}
+		} else if (scalar_type->kind() == Scalar_kind::SIGNED) {
+			if (scalar_type->buffersize() == 1L) {
+				json_data = reference.scalar_value<int8_t>();
+			} else if (scalar_type->buffersize() == 2L) {
+				json_data = reference.scalar_value<int16_t>();
+			} else if (scalar_type->buffersize() == 4L) {
+				json_data = reference.scalar_value<int32_t>();
+			} else if (scalar_type->buffersize() == 8L) {
+				json_data = reference.scalar_value<int64_t>();
+			} else {
+				throw Type_error{"Unknown size of signed integer datatype"};
+			}
+		} else if (scalar_type->kind() == Scalar_kind::FLOAT) {
+			if (scalar_type->buffersize() == 4L) {
+				json_data = reference.scalar_value<float>();
+			} else if (scalar_type->buffersize() == 8L) {
+				json_data = reference.scalar_value<double>();
+			} else {
+				throw Type_error{"Unknown size of float datatype"};
+			}
+		} else {
+			throw Type_error{"Unknown datatype to get value"};
+		}
+	}
+
+	/** Write to json an array data
+	 *
+	 * \param json_data A json data to which we write an array data
+	 * \param reference A reference to an array datatype
+	 * \param logger PDI's logger instance
+	 */
+	void write_array_to_json(nlohmann::json& json_data, Ref_r reference, Logger& logger)
+	{
+		auto array_type = std::dynamic_pointer_cast<const Array_datatype>(reference.type());
+		if (const auto&& sub_type = std::dynamic_pointer_cast<const Scalar_datatype>(array_type->subtype())) {
+			if (sub_type->kind() == Scalar_kind::UNSIGNED && sub_type->buffersize() == 1L) {
+				std::string str = "";
+				for (int i = 0; i < array_type->size(); i++) {
+					str += *reinterpret_cast<const char*>(reinterpret_cast<const char*>(Ref_r{reference[i]}.get()));
+				}
+				json_data.push_back(str);
+				return;
+			}
+		}
+		for (int i = 0; i < array_type->size(); i++) {
+			if (const auto&& sub_type = std::dynamic_pointer_cast<const Scalar_datatype>(array_type->subtype())) { // an array of scalars
+				nlohmann::json result;
+				write_scalar_to_json(result, Ref_r{reference[i]}, logger);
+				json_data.push_back(result);
```

Comment:
```suggestion
				json_data.push_back(write_scalar_to_json(Ref_r{reference[i]}, logger););
```

Reply:
I believe `write_scalar_to_json` may simply return a `nlohmann::json` result type. This would be cleaner.

Reply:
My previous comment also applies to the other array subtypes below.